PR: https://github.com/odoo/odoo/pull/233654

From: 585cccc278ef3b7eb37660dc1d731f0f80a45002
From: Zelong Lin
Date: 2025-11-04 19:18:47

Breaking data model changes scores: del:2 + add:0, change matches:
-    _inherit = ["mail.thread"]
-    _inherit = ["bus.listener.mixin"]

Total Changes: 205

[IMP] mail: introduce BusSyncMixin

Currently, Discuss Channel, Discuss Channel Member, Discuss Category,
have certain fields that need to stay synchronized when updated.

This is currently done with a custom implementation of _sync_field_names
on each model and orm (write) overrides.

Since the synchronization logic is mostly similar across these models,
we want to introduce a way to generalize this solution.

task-5221092

closes odoo/odoo#233654

Signed-off-by: Sébastien Theys (seb) <seb@odoo.com>

================================= pseudo patch: =================================

--- a/addons/mail/models/discuss/__init__.py
+++ b/addons/mail/models/discuss/__init__.py
@@ -4,6 +4,7 @@
 from . import mail_message
 
 # discuss
+from . import bus_sync_mixin
 from . import discuss_call_history
 from . import discuss_category
 from . import discuss_channel_member

--- a/None
+++ b/addons/mail/models/discuss/bus_sync_mixin.py
@@ -0,0 +1,76 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from collections import defaultdict
+from odoo import models, api
+from odoo.addons.web.models.models import lazymapping
+from odoo.addons.mail.tools.discuss import Store
+
+
+class BusSyncMixin(models.AbstractModel):
+    _name = "bus.sync.mixin"
+    _description = "Mixin for Bus Sync"
+
+    @api.model
+    def _sync_field_names(self):
+        """
+        Return the field names to sync. Override in specific models.
+        Keys are bus subchannel names, values are lists of field names to sync.
+        """
+        return defaultdict(list)
+
+    @api.model
+    def _sync_extra_field_names(self, target: Store.Target, fields):
+        """
+        Return the extra field names to sync. Override in specific models.
+        :param target: the target to sync to (Store.Target)
+        :param fields: list of field names that have changed
+        """
+        return []
+
+    def write(self, vals):
+        def get_field_name(field_description):
+            if isinstance(field_description, Store.Attr):
+                return field_description.field_name
+            return field_description
+
+        def get_field_value(record, field_description):
+            """Get the value of a field based on its description."""
+            if isinstance(field_description, Store.Attr):
+                if field_description.predicate and not field_description.predicate(record):
+                    return None
+            if isinstance(field_description, Store.Relation):
+                return field_description._get_value(record).records
+            if isinstance(field_description, Store.Attr):
+                return field_description._get_value(record)
+            return record[field_description]
+
+        def get_vals(record):
+            """Get the current values of the fields to sync."""
+            return {
+                subchannel: {
+                    get_field_name(field_description): (
+                        get_field_value(record, field_description),
+                        field_description,
+                    )
+                    for field_description in field_descriptions
+                }
+                for subchannel, field_descriptions in self._sync_field_names().items()
+            }
+
+        old_vals = {record: get_vals(record) for record in self}
+        result = super().write(vals)
+        stores = lazymapping(lambda param: Store(bus_channel=param[0], bus_subchannel=param[1]))
+        for record in self:
+            for subchannel, values in get_vals(record).items():
+                diff = []
+                for field_name, (value, field_description) in values.items():
+                    if value != old_vals[record][subchannel][field_name][0]:
+                        diff.append(field_description)
+                if diff:
+                    for channel in record._bus_channel():
+                        stores[channel, subchannel].add(record, diff)
+                        if extra_fields := record._sync_extra_field_names(stores[channel, subchannel].target, diff):
+                            stores[channel, subchannel].add(record, extra_fields)
+        for store in stores.values():
+            store.bus_send()
+        return result

--- a/addons/mail/models/discuss/discuss_category.py
+++ b/addons/mail/models/discuss/discuss_category.py
@@ -9,6 +9,7 @@ from odoo.addons.mail.tools.discuss import Store
 class DiscussCategory(models.Model):
     _name = "discuss.category"
     _description = "Discussion Category"
+    _inherit = ["bus.sync.mixin"]
 
     # description
     name = fields.Char("Name", required=True)
@@ -17,20 +18,13 @@ class DiscussCategory(models.Model):
     # constraints
     _name_unique = models.Constraint("UNIQUE(name)", "The category name must be unique")
 
-    def write(self, vals):
-        if "name" in vals:
-            old_name_by_category = {category: category.name for category in self}
-            result = super().write(vals)
-            stores = lazymapping(lambda channel: Store(bus_channel=channel))
-            for category in self:
-                if old_name_by_category[category] == vals["name"]:
-                    continue
-                for channel in category.channel_ids:
-                    stores[channel].add(category, ["name"])
-            for store in stores.values():
-                store.bus_send()
-            return result
-        return super().write(vals)
+    def _sync_field_names(self):
+        res = super()._sync_field_names()
+        res[None] += ["name"]
+        return res
+
+    def _bus_channel(self):
+        return self.channel_ids
 
     @api.ondelete(at_uninstall=False)
     def _unlink_sync_to_channel(self):

--- a/addons/mail/models/discuss/discuss_channel.py
+++ b/addons/mail/models/discuss/discuss_channel.py
@@ -52,7 +52,7 @@ class DiscussChannel(models.Model):
     _description = 'Discussion Channel'
     _mail_flat_thread = False
     _mail_post_access = 'read'
-    _inherit = ["mail.thread"]
+    _inherit = ["mail.thread", "bus.sync.mixin"]
 
     MAX_BOUNCE_LIMIT = 10
 
@@ -457,55 +457,14 @@ class DiscussChannel(models.Model):
                         channels=failing_channels.mapped("name"),
                     )
                 )
-
-        def get_field_name(field_description):
-            if isinstance(field_description, Store.Attr):
-                return field_description.field_name
-            return field_description
-
-        def get_field_value(channel, field_description):
-            if isinstance(field_description, Store.Attr):
-                if field_description.predicate and not field_description.predicate(channel):
-                    return None
-            if isinstance(field_description, Store.Relation):
-                return field_description._get_value(channel).records
-            if isinstance(field_description, Store.Attr):
-                return field_description._get_value(channel)
-            return channel[field_description]
-
-        def get_vals(channel):
-            return {
-                subchannel: {
-                    get_field_name(field_description): (
-                        get_field_value(channel, field_description),
-                        field_description,
-                    )
-                    for field_description in field_descriptions
-                }
-                for subchannel, field_descriptions in self._sync_field_names().items()
-            }
-
-        old_vals = {channel: get_vals(channel) for channel in self}
         result = super().write(vals)
-        for channel in self:
-            new_subchannel_vals = get_vals(channel)
-            for subchannel, values in new_subchannel_vals.items():
-                diff = []
-                for field_name, (value, field_description) in values.items():
-                    if value != old_vals[channel][subchannel][field_name][0]:
-                        diff.append(field_description)
-                if diff:
-                    Store(
-                        bus_channel=channel,
-                        bus_subchannel=subchannel,
-                    ).add(channel, diff).bus_send()
         if vals.get('group_ids'):
             self._subscribe_users_automatically()
         return result
 
     def _sync_field_names(self):
         # keys are bus subchannel names, values are lists of field names to sync
-        res = defaultdict(list)
+        res = super()._sync_field_names()
         res[None] += [
             Store.Attr("avatar_cache_key", predicate=is_channel_or_group),
             Store.One("discuss_category_id", ["name"], sudo=True),  # sudo: discuss.category - reading name is acceptable

--- a/addons/mail/models/discuss/discuss_channel_member.py
+++ b/addons/mail/models/discuss/discuss_channel_member.py
@@ -21,7 +21,7 @@ SFU_MODE_THRESHOLD = 3
 
 class DiscussChannelMember(models.Model):
     _name = 'discuss.channel.member'
-    _inherit = ["bus.listener.mixin"]
+    _inherit = ["bus.listener.mixin", "bus.sync.mixin"]
     _description = "Channel Member"
     _rec_names_search = ["channel_id", "partner_id", "guest_id"]
     _bypass_create_check = {}
@@ -233,48 +233,12 @@ class DiscussChannelMember(models.Model):
             for field_name in ['channel_id', 'partner_id', 'guest_id']:
                 if field_name in vals and vals[field_name] != channel_member[field_name].id:
                     raise AccessError(_('You can not write on %(field_name)s.', field_name=field_name))
-
-        def get_field_name(field_description):
-            if isinstance(field_description, Store.Attr):
-                return field_description.field_name
-            return field_description
-
-        def get_vals(member):
-            return {
-                get_field_name(field_description): (
-                    member[get_field_name(field_description)],
-                    field_description,
-                )
-                for field_description in self._sync_field_names()
-            }
-
-        old_vals_by_member = {member: get_vals(member) for member in self}
-        result = super().write(vals)
-        for member in self:
-            new_values = get_vals(member)
-            diff = []
-            for field_name, (new_value, field_description) in new_values.items():
-                old_value = old_vals_by_member[member][field_name][0]
-                if new_value != old_value:
-                    diff.append(field_description)
-            if diff:
-                store = Store(bus_channel=member._bus_channel())
-                diff.extend(
-                    [
-                        Store.One("channel_id", [], as_thread=True),
-                        *self.env["discuss.channel.member"]._to_store_persona(store.target, []),
-                    ]
-                )
-                if "message_unread_counter" in diff:
-                    # sudo: bus.bus: reading non-sensitive last id
-                    bus_last_id = self.env["bus.bus"].sudo()._bus_last_id()
-                    diff.append({"message_unread_counter_bus_id": bus_last_id})
-                store.add(member, diff).bus_send()
-        return result
+        return super().write(vals)
 
     @api.model
     def _sync_field_names(self):
-        return [
+        res = super()._sync_field_names()
+        res[None] += [
             "custom_channel_name",
             "custom_notifications",
             "is_favorite",
@@ -290,6 +254,23 @@ class DiscussChannelMember(models.Model):
             ),
             "unpin_dt",
         ]
+        return res
+
+    @api.model
+    def _sync_extra_field_names(self, target: Store.Target, fields):
+        res = super()._sync_extra_field_names(target, fields)
+        res += [
+            Store.One("channel_id", [], as_thread=True),
+            *self.env["discuss.channel.member"]._to_store_persona(target, []),
+        ]
+        if "message_unread_counter" in fields:
+            res += [
+                Store.Attr(
+                    "message_unread_counter_bus_id",
+                    value=lambda m: self.env["bus.bus"].sudo()._bus_last_id(),
+                )
+            ]
+        return res
 
     def unlink(self):
         # sudo: discuss.channel.rtc.session - cascade unlink of sessions for self member
